server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java

```diff
@@ -0,0 +1,372 @@
+        } else if (value.getClass() == String.class || (missingBucket && fieldType == null)) {
+            // the value might be not string if this field is missing in this shard but present in other shards
+            // and doesn't have a string type
+            afterValue = format.parseBytesRef(value.toString());
+            afterValueOrd = null;
+        } else {
+            throw new IllegalArgumentException("invalid value, expected string, got " + value.getClass().getSimpleName());
+        }
+        assert invariant();
+    }
+
+    @Override
+    BytesRef toComparable(int slot) throws IOException {
+        assert slot < numSlots;
+        long ord = valuesOrd.get(slot);
+        if (ord == Long.MIN_VALUE) {
+            assert missingBucket;
+            return null;
+        } else if (ord < 0) {
+            return valuesUnmapped.get(slot);
+        } else if (lastLookupOrd != null && ord == lastLookupOrd) {
+            assert ord >= 0;
+            return lastLookupValue;
+        } else {
+            assert ord >= 0;
+            lastLookupOrd = ord;
+            return lastLookupValue = BytesRef.deepCopyOf(lookup.lookupOrd(ord));
+        }
+    }
+
+    @Override
+    LeafBucketCollector getLeafCollector(LeafReaderContext context, LeafBucketCollector next) throws IOException {
+        assert invariant();
+        final SortedSetDocValues dvs = docValuesFunc.apply(context);
+        remapOrdinals(lookup, dvs);
+        lookup = dvs;
+        assert invariant();
+        return new LeafBucketCollector() {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                // caller of getLeafCollector ensures that collection happens before requesting a new leaf collector
+                // this is important as ordinals only make sense in the context of the current lookup
+                assert dvs == lookup;
+                if (dvs.advanceExact(doc)) {
+                    long ord;
+                    while ((ord = dvs.nextOrd()) != NO_MORE_ORDS) {
+                        currentValueOrd = ord;
+                        currentValueUnmapped = null;
+                        next.collect(doc, bucket);
+                    }
+                } else if (missingBucket) {
+                    currentValueOrd = Long.MIN_VALUE;
+                    currentValueUnmapped = null;
+                    next.collect(doc, bucket);
+                }
+            }
+        };
+    }
+
+    @Override
+    LeafBucketCollector getLeafCollector(Comparable value, LeafReaderContext context, LeafBucketCollector next) throws IOException {
+        assert invariant();
+        if (value.getClass() != BytesRef.class) {
+            throw new IllegalArgumentException("Expected BytesRef, got " + value.getClass());
+        }
+        BytesRef term = (BytesRef) value;
+        final SortedSetDocValues dvs = docValuesFunc.apply(context);
+        remapOrdinals(lookup, dvs);
```

Comment:
We call this function once per bucket in SortedDocsProducer so we don't need to remap the ordinals each time.
We should keep track of the context ord and remap only if the provided context ord is different ?

Reply:
good idea. addressed in 64b1872